hazelcast/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java

```diff
@@ -1102,15 +1108,20 @@ private String dagToJson(DAG dag) {
     ) {
         // the order of operations is important.
         long jobId = jobRecord.getJobId();
-        JobResult jobResult = jobRepository.getJobResult(jobId);
-        if (jobResult != null) {
-            logger.fine("Not starting job " + idToString(jobId) + ", already has result: " + jobResult);
-            return jobResult.asCompletableFuture();
-        }
 
         MasterContext masterContext;
         MasterContext oldMasterContext;
         synchronized (lock) {
+            // We check the JobResult while holding the lock to avoid this scenario:
+            // 1. We find no job result
+            // 2. Another thread creates the result and removes the master context in completeJob
+            // 3. We re-create the master context below
+            JobResult jobResult = jobRepository.getJobResult(jobId);
```

Comment:
I think we can remove the `completeMasterContextIfJobAlreadyCompleted` call below. Now we ensure that if JobResult exists, we never create a new master context, so we don't need to double-check. If we don't fully ensure this, we still have the issue.

Maybe we can replace the call with an assert to make sure that JobResult doesn't exist.

Reply:
After removing this `completeMasterContextIfJobAlreadyCompleted` call below, the topology change tests with disabled autoscaling: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/test/java/com/hazelcast/jet/core/TopologyChangeTest.java#L211-L219, started to fail. See: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java#L1148-L1153